[WPE] Import the initial painting implementation using Skia

carlosgcampos opened the pull request:
#### de802f3b8dfb14d2f8ca791924761285726ab392
<pre>
[WPE] Import the initial painting implementation using Skia
<a href="https://bugs.webkit.org/show_bug.cgi?id=268975">https://bugs.webkit.org/show_bug.cgi?id=268975</a>

Reviewed by Don Olmstead.

Add initial Skia implemention for the WPE port.

Co-authored-by: Adrian Perez de Castro &lt;aperez@igalia.com&gt;
Co-authored-by: Alejandro G. Castro &lt;alex@igalia.com&gt;
Co-authored-by: Nikolas Zimmermann &lt;nzimmermann@igalia.com&gt;

* Source/CMakeLists.txt:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::shouldAccelerate const):
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/ImageDecoders.cmake:
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::DataSegment::create):
(WebCore::DataSegment::data const):
(WebCore::DataSegment::size const):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::DataSegment::DataSegment):
* Source/WebCore/platform/Skia.cmake: Added.
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/SourcesSkia.txt: Added.
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/FloatRect.h:
* Source/WebCore/platform/graphics/Font.cpp:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::fontManager const):
(WebCore::FontCache::harfBuzzFontCache):
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::skFont const):
(WebCore::FontPlatformData::skiaHarfBuzzFont const):
(WebCore::FontPlatformData::features const):
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/GraphicsContextGLImageExtractor.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/ImageBufferPlatformBackend.h:
* Source/WebCore/platform/graphics/Pattern.h:
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::skiaGrContext):
* Source/WebCore/platform/graphics/PlatformGraphicsContext.h:
* Source/WebCore/platform/graphics/PlatformImage.h:
* Source/WebCore/platform/graphics/PlatformPath.h:
* Source/WebCore/platform/graphics/PlatformPathImpl.h:
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::operator==):
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerCairo.cpp:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp: Added.
(WebCore::ImageGStreamer::ImageGStreamer):
(WebCore::ImageGStreamer::~ImageGStreamer):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::fromNativeImage):
* Source/WebCore/platform/graphics/harfbuzz/HbUniquePtr.h:
(WebCore::HbPtrDeleter&lt;hb_blob_t&gt;::operator() const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBackingStore.cpp:
(Nicosia::BackingStore::updateTile):
* Source/WebCore/platform/graphics/nicosia/NicosiaBackingStore.h:
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::create):
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::~UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::beginPainting):
(Nicosia::UnacceleratedBuffer::completePainting):
(Nicosia::UnacceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::AcceleratedBuffer::create):
(Nicosia::AcceleratedBuffer::AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::size const):
(Nicosia::AcceleratedBuffer::beginPainting):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::Buffer::Buffer):
(Nicosia::Buffer::create): Deleted.
(Nicosia::Buffer::~Buffer): Deleted.
(Nicosia::Buffer::beginPainting): Deleted.
(Nicosia::Buffer::completePainting): Deleted.
(Nicosia::Buffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
(Nicosia::Buffer::platformWaitUntilPaintingComplete):
(Nicosia::Buffer::supportsAlpha const):
(Nicosia::Buffer::surface const):
(Nicosia::Buffer::size const): Deleted.
(Nicosia::Buffer::stride const): Deleted.
(Nicosia::Buffer::data const): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaPaintingContextCairo.cpp:
(Nicosia::PaintingContextCairo::ForPainting::ForPainting):
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp: Added.
(WebCore::harfBuzzPositionToFloat):
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
(WebCore::characterScript):
(WebCore::findNextRun):
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
* Source/WebCore/platform/graphics/skia/DrawGlyphsRecorderSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::DrawGlyphsRecorder::DrawGlyphsRecorder):
(WebCore::DrawGlyphsRecorder::drawGlyphs):
* Source/WebCore/platform/graphics/skia/FloatRectSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::FloatRect::FloatRect):
(WebCore::FloatRect::operator SkRect const):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp: Added.
(WebCore::FontCache::platformInit):
(WebCore::FontCache::systemFallbackForCharacterCluster):
(WebCore::FontCache::systemFontFamilies):
(WebCore::FontCache::isSystemFontForbiddenForEditing):
(WebCore::FontCache::lastResortFallbackFont):
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily):
(WebCore::getFamilyNameStringFromFamily):
(WebCore::FontCache::computeFeatures):
(WebCore::FontCache::createFontPlatformData):
(WebCore::FontCache::platformAlternateFamilyName):
(WebCore::FontCache::platformInvalidate):
(WebCore::FontCache::platformPurgeInactiveFontData):
* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp: Added.
(WebCore::FontCascade::drawGlyphs):
(WebCore::FontCascade::canReturnFallbackFontsForComplexText):
(WebCore::FontCascade::canExpandAroundIdeographsInComplexText):
(WebCore::FontCascade::resolveEmojiPolicy):
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp: Added.
(WebCore::FontCustomPlatformData::FontCustomPlatformData):
(WebCore::FontCustomPlatformData::fontPlatformData):
(WebCore::createFontCustomPlatformData):
(WebCore::FontCustomPlatformData::supportsFormat):
(WebCore::FontCustomPlatformData::supportsTechnology):
* Source/WebCore/platform/graphics/skia/FontDescriptionSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::FontCascadeDescription::effectiveFamilyCount const):
(WebCore::FontCascadeDescription::effectiveFamilyAt const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp: Added.
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::isFixedPitch const):
(WebCore::FontPlatformData::hash const):
(WebCore::FontPlatformData::platformIsEqual const):
(WebCore::FontPlatformData::description const):
(WebCore::FontPlatformData::familyName const):
(WebCore::FontPlatformData::openTypeTable const):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
(WebCore::FontPlatformData::hbFont const):
(WebCore::FontPlatformData::createOpenTypeMathHarfBuzzFont const):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::variationAxes const):
* Source/WebCore/platform/graphics/skia/FontSkia.cpp: Added.
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::Font::determinePitch):
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::platformSupportsCodePoint const):
* Source/WebCore/platform/graphics/skia/FontVariationsSkia.cpp: Added.
(WebCore::defaultFontVariationValues):
* Source/WebCore/platform/graphics/skia/FontVariationsSkia.h: Added.
(WebCore::FontVariationDefaults::clamp const):
* Source/WebCore/platform/graphics/skia/GlyphPageSkia.cpp: Added.
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/skia/GradientSkia.cpp: Added.
(WebCore::Gradient::stopsChanged):
(WebCore::webCoreDoubleToSkScalar):
(WebCore::toSkiaInterpolation):
(WebCore::Gradient::getGradient):
(WebCore::Gradient::fill):
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp: Added.
(WebCore::GraphicsContextGLImageExtractor::extractImage):
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp: Added.
(WebCore::GraphicsContextSkia::GraphicsContextSkia):
(WebCore::GraphicsContextSkia::hasPlatformContext const):
(WebCore::GraphicsContextSkia::getCTM const):
(WebCore::GraphicsContextSkia::platformContext const):
(WebCore::GraphicsContextSkia::save):
(WebCore::GraphicsContextSkia::restore):
(WebCore::GraphicsContextSkia::drawRect):
(WebCore::toSkiaBlendMode):
(WebCore::toSkSamplingOptions):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawLine):
(WebCore::GraphicsContextSkia::drawEllipse):
(WebCore::toSkiaFillType):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::strokePath):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::createStrokePaint const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::resetClip):
(WebCore::GraphicsContextSkia::clip):
(WebCore::GraphicsContextSkia::clipPath):
(WebCore::GraphicsContextSkia::clipBounds const):
(WebCore::GraphicsContextSkia::clipToImageBuffer):
(WebCore::GraphicsContextSkia::drawFocusRing):
(WebCore::GraphicsContextSkia::drawLinesForText):
(WebCore::GraphicsContextSkia::drawDotsForDocumentMarker):
(WebCore::GraphicsContextSkia::translate):
(WebCore::GraphicsContextSkia::didUpdateState):
(WebCore::GraphicsContextSkia::concatCTM):
(WebCore::GraphicsContextSkia::setCTM):
(WebCore::GraphicsContextSkia::beginTransparencyLayer):
(WebCore::GraphicsContextSkia::endTransparencyLayer):
(WebCore::GraphicsContextSkia::clearRect):
(WebCore::GraphicsContextSkia::strokeRect):
(WebCore::GraphicsContextSkia::setLineCap):
(WebCore::GraphicsContextSkia::setLineDash):
(WebCore::GraphicsContextSkia::setLineJoin):
(WebCore::GraphicsContextSkia::setMiterLimit):
(WebCore::GraphicsContextSkia::clipOut):
(WebCore::GraphicsContextSkia::rotate):
(WebCore::GraphicsContextSkia::scale):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
(WebCore::GraphicsContextSkia::drawPattern):
(WebCore::GraphicsContextSkia::renderingMode const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h: Added.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp: Added.
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaAcceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::ImageBufferSkiaBackend::debugDescription const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.h: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.cpp: Added.
(WebCore::ImageBufferSkiaSurfaceBackend::calculateSafeBackendSize):
(WebCore::ImageBufferSkiaSurfaceBackend::calculateBytesPerRow):
(WebCore::ImageBufferSkiaSurfaceBackend::calculateMemoryCost):
(WebCore::ImageBufferSkiaSurfaceBackend::ImageBufferSkiaSurfaceBackend):
(WebCore::ImageBufferSkiaSurfaceBackend::bytesPerRow const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp: Added.
(WebCore::ImageBufferSkiaUnacceleratedBackend::create):
(WebCore::ImageBufferSkiaUnacceleratedBackend::ImageBufferSkiaUnacceleratedBackend):
(WebCore::ImageBufferSkiaUnacceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaUnacceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaUnacceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaUnacceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::encodeData):
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.h: Copied from Source/WebCore/platform/graphics/PlatformPathImpl.h.
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::PlatformImageNativeImageBackend::size const):
(WebCore::PlatformImageNativeImageBackend::hasAlpha const):
(WebCore::PlatformImageNativeImageBackend::colorSpace const):
(WebCore::NativeImage::singlePixelSolidColor const):
(WebCore::NativeImage::draw):
(WebCore::NativeImage::clearSubimages):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp: Added.
(WebCore::PathSkia::create):
(WebCore::PathSkia::PathSkia):
(WebCore::PathSkia::copy const):
(WebCore::PathSkia::platformPath const):
(WebCore::PathSkia::add):
(WebCore::PathSkia::addPath):
(WebCore::PathSkia::applySegments const):
(WebCore::PathSkia::applyElements const):
(WebCore::PathSkia::isEmpty const):
(WebCore::PathSkia::currentPoint const):
(WebCore::PathSkia::transform):
(WebCore::PathSkia::contains const):
(WebCore::PathSkia::strokeContains const):
(WebCore::PathSkia::fastBoundingRect const):
(WebCore::PathSkia::boundingRect const):
(WebCore::PathSkia::strokeBoundingRect const):
* Source/WebCore/platform/graphics/skia/PathSkia.h: Added.
(isType):
* Source/WebCore/platform/graphics/skia/PatternSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::Pattern::createPlatformPattern const):
* Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::PlatformDisplay::skiaGLContext):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFont.cpp: Added.
(WebCore::SkiaHarfBuzzFont::getOrCreate):
(WebCore::harfBuzzFontFunctions):
(WebCore::createHarfBuzzFace):
(WebCore::SkiaHarfBuzzFont::SkiaHarfBuzzFont):
(WebCore::SkiaHarfBuzzFont::~SkiaHarfBuzzFont):
(WebCore::floatToHarfBuzzPosition):
(WebCore::doubleToHarfBuzzPosition):
(WebCore::SkiaHarfBuzzFont::scaledFont):
(WebCore::SkiaHarfBuzzFont::glyph):
(WebCore::SkiaHarfBuzzFont::glyphWidth):
(WebCore::SkiaHarfBuzzFont::glyphWidths):
(WebCore::SkiaHarfBuzzFont::glyphExtents):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFont.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFontCache.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::SkiaHarfBuzzFontCache::font):
(WebCore::SkiaHarfBuzzFontCache::remove):
(WebCore::SkiaHarfBuzzFontCache::clear):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFontCache.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/TransformationMatrixSkia.cpp: Added.
(WebCore::TransformationMatrix::TransformationMatrix):
(WebCore::TransformationMatrix::operator SkM44 const):
(WebCore::AffineTransform::AffineTransform):
(WebCore::AffineTransform::operator SkMatrix const):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::updateContents):
(WebCore::BitmapTexture::copyFromExternalTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawNumber):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerCairo.cpp: Added.
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp: Added.
(WebCore::AcceleratedBufferPool::singleton):
(WebCore::AcceleratedBufferPool::AcceleratedBufferPool):
(WebCore::AcceleratedBufferPool::~AcceleratedBufferPool):
(WebCore::AcceleratedBufferPool::acquireBuffer):
(WebCore::AcceleratedBufferPool::Entry::Entry):
(WebCore::AcceleratedBufferPool::Entry::markIsInUse):
(WebCore::AcceleratedBufferPool::Entry::canBeReleased const):
(WebCore::AcceleratedBufferPool::createAcceleratedBuffer):
(WebCore::AcceleratedBufferPool::scheduleReleaseUnusedBuffers):
(WebCore::AcceleratedBufferPool::releaseUnusedBuffersTimerFired):
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebCore/platform/graphics/texmap/coordinated/SurfaceUpdateInfo.h: Removed.
* Source/WebCore/platform/graphics/texmap/coordinated/Tile.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStoreClient.h:
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp: Copied from Source/WebCore/platform/graphics/ImageBufferPlatformBackend.h.
(WebCore::ImageBackingStore::image const):
* Source/WebCore/platform/skia/SharedBufferSkia.cpp: Added.
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create):
(WebCore::SharedBuffer::createSkData const):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/API/c/cairo/WKImageCairo.cpp:
(WKImageCreateCairoSurface):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::updateBackingStore):
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebKit/Shared/ShareableBitmap.h:
(WebKit::ShareableBitmapConfiguration::imageInfo const):
* Source/WebKit/Shared/skia/ShareableBitmapSkia.cpp: Added.
(WebKit::ShareableBitmapConfiguration::validateColorSpace):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerRow):
(WebKit::ShareableBitmap::createGraphicsContext):
(WebKit::ShareableBitmap::paint):
(WebKit::ShareableBitmap::createImage):
(WebKit::ShareableBitmap::setOwnershipOfMemory):
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp: Copied from Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.h.
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::encodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::decodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::decodePlatformData):
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::setCursor):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::IconCache::iconPath):
(WebKit::IconCache::iconBytes):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::CompositingCoordinator):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.h:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
* Source/cmake/OptionsWPE.cmake:
* Tools/ImageDiff/Cairo.cmake:
* Tools/ImageDiff/PlatformImage.h:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewWindowProperties):
* Tools/TestWebKitAPI/glib/PlatformWPE.cmake:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.h:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/WebKitTestRunner/PlatformWebView.h:
* Tools/WebKitTestRunner/cairo/TestInvocationCairo.cpp:
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClient.h:
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.cpp:
(WTR::PlatformWebViewClientLibWPE::snapshot):
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.h:
* Tools/WebKitTestRunner/skia/TestInvocationSkia.cpp: Added.
(WTR::computeSHA1HashStringForPixmap):
(WTR::dumpPixmap):
(WTR::TestInvocation::dumpPixelsAndCompareWithExpected):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::snapshot):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.h:
* Tools/WebKitTestRunner/wpe/TestControllerWPE.cpp:
(WTR::TestController::takeViewPortSnapshot):
* Tools/wpe/backends/CMakeLists.txt:
* Tools/wpe/backends/HeadlessViewBackend.h:
* Tools/wpe/backends/fdo/HeadlessViewBackendFdo.cpp:
(WPEToolingBackends::HeadlessViewBackend::HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::~HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::snapshot):
(WPEToolingBackends::HeadlessViewBackend::updateSnapshot):

Canonical link: <a href="https://commits.webkit.org/274457@main">https://commits.webkit.org/274457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62f6225985b46b5e883ea7226b95052cf8f1d851

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41705 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34888 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41477 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/21008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15479 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32783 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15269 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13263 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13231 "layout-tests (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42982 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/32624 "JSC test binary failure: testapi") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35220 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/38797 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13983 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11537 "9 flakes 1 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37279 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15589 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34169 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/45803 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15252 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9331 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5114 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15075 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -234,12 +235,13 @@ if (ENABLE_SPEECH_SYNTHESIS)
+elseif (USE_SKIA)
+    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO FALSE)
 endif ()
 
```
You have to flip this around, at this point `USE_CAIRO` is not yet defined, only `USE_SKIA`. The options are mutually exclusive so the following will do:

```suggestion
if (USE_SKIA)
    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO FALSE)
else ()
    find_package(Cairo 1.16.0 REQUIRED)
    find_package(Fontconfig 2.13.0 REQUIRED)
    find_package(Freetype 2.9.0 REQUIRED)
    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO TRUE)
endif ()

```

carlosgcampos replied:
Thanks!

aperezdc commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -80,6 +84,18 @@ RefPtr<ImageBuffer> ImageBuffer::create(const FloatSize& size, RenderingPurpose
 #endif
 
+#if USE(SKIA)
+    static const char* enableCPURendering = getenv("WEBKIT_SKIA_ENABLE_CPU_RENDERING");
+    if (!enableCPURendering || !strcmp(enableCPURendering, "0")) {
+        static const char* disableAccelerated2DCanvas = getenv("WEBKIT_DISABLE_ACCELERATED_2D_CANVAS");
```
It would be nicer to have these as runtime feature flags, but we can handle that in a separate patch... Or “patches” in plural, because at some point it would be nice to reduce the number of environment variables in favor of feature flags 😉

carlosgcampos replied:
Yes, the env vars introduced in this patch are just for development, I don't know if they will be removed, replaced with a setting or runtime feature, or what.

donny-dont commented on Source/CMakeLists.txt:
```diff
@@ -30,6 +30,10 @@ if (USE_LIBWEBRTC)
     include_directories(${CMAKE_CURRENT_BINARY_DIR}/ThirdParty/libwebrtc)
 endif ()
 
+if (USE_SKIA)
+    add_subdirectory(ThirdParty/skia)
+endif ()
```
Just making a note that this is after ANGLE which is good since Windows will want that to be present.

donny-dont commented on Source/WebCore/PlatformWPE.cmake:
```diff
@@ -26,6 +32,7 @@ list(APPEND WebCore_PRIVATE_INCLUDE_DIRECTORIES
     "${WEBCORE_DIR}/platform/graphics/epoxy"
     "${WEBCORE_DIR}/platform/graphics/gbm"
     "${WEBCORE_DIR}/platform/graphics/gstreamer"
+    "${WEBCORE_DIR}/platform/graphics/harfbuzz"
```
Remove its in Skia.cmake

donny-dont commented on Source/WebCore/PlatformWPE.cmake:
```diff
@@ -58,6 +65,7 @@ set(CSS_VALUE_PLATFORM_DEFINES "HAVE_OS_DARK_MODE_SUPPORT=1")
 set(WebCore_USER_AGENT_SCRIPTS_DEPENDENCIES ${WEBCORE_DIR}/platform/wpe/RenderThemeWPE.cpp)
 
 list(APPEND WebCore_LIBRARIES
+    HarfBuzz::HarfBuzz
```
Move into Skia.cmake

donny-dont commented on Source/WebCore/platform/graphics/FontCustomPlatformData.h:
```diff
@@ -75,8 +77,10 @@ struct FontCustomPlatformData : public RefCounted<FontCustomPlatformData> {
+#elif USE(CAIRO)
     FontCustomPlatformData(FT_Face, FontPlatformData::CreationData&&);
+#elif USE(SKIA)
+    FontCustomPlatformData(sk_sp<SkTypeface>&&, FontPlatformData::CreationData&&);
```
Not asking for a solution right this second but in order to use the GPU Process we're going to have to have a way to serialize this information.

aperezdc replied:
There's methods `SkTypeface::serialize()` and a static `SkTypeface::MakeDeserialize()`  factory function to re-create a new instance from the serialized data; probably these will come in handy here.

donny-dont commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -191,8 +199,10 @@ class FontPlatformData {
+#if USE(FREETYPE) && USE(CAIRO)
     FontPlatformData(cairo_font_face_t*, RefPtr<FcPattern>&&, float size, bool fixedWidth, bool syntheticBold, bool syntheticOblique, FontOrientation, const FontCustomPlatformData* = nullptr);
+#elif USE(SKIA)
+    FontPlatformData(sk_sp<SkTypeface>&&, float size, bool syntheticBold, bool syntheticOblique, FontOrientation, FontWidthVariant, TextRenderingMode, Vector<hb_feature_t>&&, const FontCustomPlatformData* = nullptr);
```
Again be aware that we're going to need to serialize this in GPU Process land.

donny-dont commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -80,26 +98,93 @@ Buffer::~Buffer()
+    auto* grContext = WebCore::PlatformDisplay::sharedDisplayForCompositing().skiaGrContext();
+    // FIXME: check if glFence is available and if not pass GrSyncCpu::kYes.
+    grContext->flushAndSubmit(m_surface.get(), GrSyncCpu::kNo);
+    m_sync = glFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
```
Not a fan of gl calls hanging out here. Is there anything in the context that would do the same?

carlosgcampos replied:
Skia supports fences internally, but I didn't find any public API to use them externally.

donny-dont commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -80,26 +98,93 @@ Buffer::~Buffer()
+void AcceleratedBuffer::waitUntilPaintingComplete()
+{
+    if (m_sync)
+        glClientWaitSync(m_sync, 0, GL_TIMEOUT_IGNORED);
```
Same. Anything in the context that would do this?

donny-dont commented on Source/WebKit/PlatformWPE.cmake:
```diff
@@ -418,8 +418,6 @@ list(APPEND WebKit_SYSTEM_INCLUDE_DIRECTORIES
     ATK::Bridge
-    Cairo::Cairo
-    Freetype::Freetype
     HarfBuzz::HarfBuzz
```
I think all the Harfbuzz code is in WebCore so you can probably remove this too

donny-dont commented on Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
```diff
@@ -0,0 +1,59 @@
+
+namespace IPC {
+
+void ArgumentCoder<WebCore::Font>::encodePlatformData(Encoder&, const WebCore::Font&)
+{
+    ASSERT_NOT_REACHED();
+}
+
+std::optional<WebCore::FontPlatformData> ArgumentCoder<WebCore::Font>::decodePlatformData(Decoder&)
+{
+    ASSERT_NOT_REACHED();
+    return std::nullopt;
+}
+
+void ArgumentCoder<WebCore::FontPlatformData::Attributes>::encodePlatformData(Encoder&, const WebCore::FontPlatformData::Attributes&)
+{
+    ASSERT_NOT_REACHED();
+}
+
+bool ArgumentCoder<WebCore::FontPlatformData::Attributes>::decodePlatformData(Decoder&, WebCore::FontPlatformData::Attributes&)
+{
+    ASSERT_NOT_REACHED();
+    return false;
+}
```
We're going to want to generate these in the future. At the moment this is fine because its unused until you add the GPU Process

donny-dont commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -234,12 +235,13 @@ if (ENABLE_SPEECH_SYNTHESIS)
-SET_AND_EXPOSE_TO_BUILD(USE_CAIRO TRUE)
-
-if (USE_CAIRO)
+if (USE_SKIA)
+    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO FALSE)
+else ()
     find_package(Cairo 1.16.0 REQUIRED)
     find_package(Fontconfig 2.13.0 REQUIRED)
     find_package(Freetype 2.9.0 REQUIRED)
+    SET_AND_EXPOSE_TO_BUILD(USE_CAIRO TRUE)
```
Since the dependencies are the same between Cairo and Skia I would probably just make this an unconditional check.

I know you've added the check into the Skia CMake but I'm not sure how well that's going to play on other platforms and honestly I think failing earlier is better.

donny-dont commented on Tools/ImageDiff/Cairo.cmake:
```diff
@@ -2,6 +2,10 @@ list(APPEND ImageDiff_SOURCES
     cairo/PlatformImageCairo.cpp
 )
 
+if (NOT TARGET Cairo::Cairo)
+    find_package(Cairo REQUIRED)
+endif ()
```
Would this ever happen? Seems wrong.

aperezdc replied:
ImageDiff does not have a Skia implementation at the moment, so with `DEVELOPER_MODE=ON` and `USE_SKIA=ON` we reach here without the `Cairo::Cairo` target being defined. So we bring it into scope if needed. One reason to do it lazily here is to avoid checking for Cairo at all in “normal” (non-developer) builds.

donny-dont replied:
Got it. Would be nice to stub out a `Skia.cmake` here but wouldn't block on that.

donny-dont commented on Tools/TestWebKitAPI/PlatformWPE.cmake:
```diff
@@ -51,6 +51,11 @@ list(APPEND TestWebCore_SYSTEM_INCLUDE_DIRECTORIES
     ${LIBSOUP_INCLUDE_DIRS}
 )
 
+list(APPEND TestWebCore_LIBRARIES
+    HarfBuzz::HarfBuzz
+    HarfBuzz::ICU
+)
```
Feel like this dependency would've ended up here already.

donny-dont reviewed (approved):
Overall happy with the patch. The build parts look good. Please add a `--skia` flag to `build-webkit` so it can be used to set `USE_SKIA` in the CMake.

If there's anyone with more experience using Skia they should prolly take a deeper look on the actual implementation and how the API is used.

carlosgcampos committed:
[WPE] Import the initial painting implementation using Skia
https://bugs.webkit.org/show_bug.cgi?id=268975

Reviewed by Don Olmstead.

Add initial Skia implemention for the WPE port.

Co-authored-by: Adrian Perez de Castro <aperez@igalia.com>
Co-authored-by: Alejandro G. Castro <alex@igalia.com>
Co-authored-by: Nikolas Zimmermann <nzimmermann@igalia.com>

* Source/CMakeLists.txt:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::shouldAccelerate const):
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/ImageDecoders.cmake:
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::DataSegment::create):
(WebCore::DataSegment::data const):
(WebCore::DataSegment::size const):
* Source/WebCore/platform/SharedBuffer.h:
(WebCore::DataSegment::DataSegment):
* Source/WebCore/platform/Skia.cmake: Added.
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/SourcesSkia.txt: Added.
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/FloatRect.h:
* Source/WebCore/platform/graphics/Font.cpp:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/FontCache.h:
(WebCore::FontCache::fontManager const):
(WebCore::FontCache::harfBuzzFontCache):
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::skFont const):
(WebCore::FontPlatformData::skiaHarfBuzzFont const):
(WebCore::FontPlatformData::features const):
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/GraphicsContextGLImageExtractor.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/ImageBufferPlatformBackend.h:
* Source/WebCore/platform/graphics/Pattern.h:
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::skiaGrContext):
* Source/WebCore/platform/graphics/PlatformGraphicsContext.h:
* Source/WebCore/platform/graphics/PlatformImage.h:
* Source/WebCore/platform/graphics/PlatformPath.h:
* Source/WebCore/platform/graphics/PlatformPathImpl.h:
* Source/WebCore/platform/graphics/WidthCache.h:
(WebCore::operator==):
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerCairo.cpp:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp: Added.
(WebCore::ImageGStreamer::ImageGStreamer):
(WebCore::ImageGStreamer::~ImageGStreamer):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::fromNativeImage):
* Source/WebCore/platform/graphics/harfbuzz/HbUniquePtr.h:
(WebCore::HbPtrDeleter<hb_blob_t>::operator() const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBackingStore.cpp:
(Nicosia::BackingStore::updateTile):
* Source/WebCore/platform/graphics/nicosia/NicosiaBackingStore.h:
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::create):
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::~UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::beginPainting):
(Nicosia::UnacceleratedBuffer::completePainting):
(Nicosia::UnacceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::AcceleratedBuffer::create):
(Nicosia::AcceleratedBuffer::AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::size const):
(Nicosia::AcceleratedBuffer::beginPainting):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::waitUntilPaintingComplete):
(Nicosia::Buffer::Buffer):
(Nicosia::Buffer::create): Deleted.
(Nicosia::Buffer::~Buffer): Deleted.
(Nicosia::Buffer::beginPainting): Deleted.
(Nicosia::Buffer::completePainting): Deleted.
(Nicosia::Buffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
(Nicosia::Buffer::platformWaitUntilPaintingComplete):
(Nicosia::Buffer::supportsAlpha const):
(Nicosia::Buffer::surface const):
(Nicosia::Buffer::size const): Deleted.
(Nicosia::Buffer::stride const): Deleted.
(Nicosia::Buffer::data const): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaPaintingContextCairo.cpp:
(Nicosia::PaintingContextCairo::ForPainting::ForPainting):
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp: Added.
(WebCore::harfBuzzPositionToFloat):
(WebCore::ComplexTextController::ComplexTextRun::ComplexTextRun):
(WebCore::characterScript):
(WebCore::findNextRun):
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
* Source/WebCore/platform/graphics/skia/DrawGlyphsRecorderSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::DrawGlyphsRecorder::DrawGlyphsRecorder):
(WebCore::DrawGlyphsRecorder::drawGlyphs):
* Source/WebCore/platform/graphics/skia/FloatRectSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::FloatRect::FloatRect):
(WebCore::FloatRect::operator SkRect const):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp: Added.
(WebCore::FontCache::platformInit):
(WebCore::FontCache::systemFallbackForCharacterCluster):
(WebCore::FontCache::systemFontFamilies):
(WebCore::FontCache::isSystemFontForbiddenForEditing):
(WebCore::FontCache::lastResortFallbackFont):
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily):
(WebCore::getFamilyNameStringFromFamily):
(WebCore::FontCache::computeFeatures):
(WebCore::FontCache::createFontPlatformData):
(WebCore::FontCache::platformAlternateFamilyName):
(WebCore::FontCache::platformInvalidate):
(WebCore::FontCache::platformPurgeInactiveFontData):
* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp: Added.
(WebCore::FontCascade::drawGlyphs):
(WebCore::FontCascade::canReturnFallbackFontsForComplexText):
(WebCore::FontCascade::canExpandAroundIdeographsInComplexText):
(WebCore::FontCascade::resolveEmojiPolicy):
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp: Added.
(WebCore::FontCustomPlatformData::FontCustomPlatformData):
(WebCore::FontCustomPlatformData::fontPlatformData):
(WebCore::createFontCustomPlatformData):
(WebCore::FontCustomPlatformData::supportsFormat):
(WebCore::FontCustomPlatformData::supportsTechnology):
* Source/WebCore/platform/graphics/skia/FontDescriptionSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::FontCascadeDescription::effectiveFamilyCount const):
(WebCore::FontCascadeDescription::effectiveFamilyAt const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp: Added.
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::isFixedPitch const):
(WebCore::FontPlatformData::hash const):
(WebCore::FontPlatformData::platformIsEqual const):
(WebCore::FontPlatformData::description const):
(WebCore::FontPlatformData::familyName const):
(WebCore::FontPlatformData::openTypeTable const):
(WebCore::FontPlatformData::create):
(WebCore::FontPlatformData::attributes const):
(WebCore::FontPlatformData::hbFont const):
(WebCore::FontPlatformData::createOpenTypeMathHarfBuzzFont const):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::variationAxes const):
* Source/WebCore/platform/graphics/skia/FontSkia.cpp: Added.
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::Font::determinePitch):
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::platformSupportsCodePoint const):
* Source/WebCore/platform/graphics/skia/FontVariationsSkia.cpp: Added.
(WebCore::defaultFontVariationValues):
* Source/WebCore/platform/graphics/skia/FontVariationsSkia.h: Added.
(WebCore::FontVariationDefaults::clamp const):
* Source/WebCore/platform/graphics/skia/GlyphPageSkia.cpp: Added.
(WebCore::GlyphPage::fill):
* Source/WebCore/platform/graphics/skia/GradientSkia.cpp: Added.
(WebCore::Gradient::stopsChanged):
(WebCore::webCoreDoubleToSkScalar):
(WebCore::toSkiaInterpolation):
(WebCore::Gradient::getGradient):
(WebCore::Gradient::fill):
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp: Added.
(WebCore::GraphicsContextGLImageExtractor::extractImage):
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp: Added.
(WebCore::GraphicsContextSkia::GraphicsContextSkia):
(WebCore::GraphicsContextSkia::hasPlatformContext const):
(WebCore::GraphicsContextSkia::getCTM const):
(WebCore::GraphicsContextSkia::platformContext const):
(WebCore::GraphicsContextSkia::save):
(WebCore::GraphicsContextSkia::restore):
(WebCore::GraphicsContextSkia::drawRect):
(WebCore::toSkiaBlendMode):
(WebCore::toSkSamplingOptions):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawLine):
(WebCore::GraphicsContextSkia::drawEllipse):
(WebCore::toSkiaFillType):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::strokePath):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::createStrokePaint const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::resetClip):
(WebCore::GraphicsContextSkia::clip):
(WebCore::GraphicsContextSkia::clipPath):
(WebCore::GraphicsContextSkia::clipBounds const):
(WebCore::GraphicsContextSkia::clipToImageBuffer):
(WebCore::GraphicsContextSkia::drawFocusRing):
(WebCore::GraphicsContextSkia::drawLinesForText):
(WebCore::GraphicsContextSkia::drawDotsForDocumentMarker):
(WebCore::GraphicsContextSkia::translate):
(WebCore::GraphicsContextSkia::didUpdateState):
(WebCore::GraphicsContextSkia::concatCTM):
(WebCore::GraphicsContextSkia::setCTM):
(WebCore::GraphicsContextSkia::beginTransparencyLayer):
(WebCore::GraphicsContextSkia::endTransparencyLayer):
(WebCore::GraphicsContextSkia::clearRect):
(WebCore::GraphicsContextSkia::strokeRect):
(WebCore::GraphicsContextSkia::setLineCap):
(WebCore::GraphicsContextSkia::setLineDash):
(WebCore::GraphicsContextSkia::setLineJoin):
(WebCore::GraphicsContextSkia::setMiterLimit):
(WebCore::GraphicsContextSkia::clipOut):
(WebCore::GraphicsContextSkia::rotate):
(WebCore::GraphicsContextSkia::scale):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
(WebCore::GraphicsContextSkia::drawPattern):
(WebCore::GraphicsContextSkia::renderingMode const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h: Added.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp: Added.
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaAcceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::ImageBufferSkiaBackend::debugDescription const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaBackend.h: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.cpp: Added.
(WebCore::ImageBufferSkiaSurfaceBackend::calculateSafeBackendSize):
(WebCore::ImageBufferSkiaSurfaceBackend::calculateBytesPerRow):
(WebCore::ImageBufferSkiaSurfaceBackend::calculateMemoryCost):
(WebCore::ImageBufferSkiaSurfaceBackend::ImageBufferSkiaSurfaceBackend):
(WebCore::ImageBufferSkiaSurfaceBackend::bytesPerRow const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp: Added.
(WebCore::ImageBufferSkiaUnacceleratedBackend::create):
(WebCore::ImageBufferSkiaUnacceleratedBackend::ImageBufferSkiaUnacceleratedBackend):
(WebCore::ImageBufferSkiaUnacceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaUnacceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaUnacceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaUnacceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformImage.h.
(WebCore::encodeData):
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.h: Copied from Source/WebCore/platform/graphics/PlatformPathImpl.h.
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::PlatformImageNativeImageBackend::size const):
(WebCore::PlatformImageNativeImageBackend::hasAlpha const):
(WebCore::PlatformImageNativeImageBackend::colorSpace const):
(WebCore::NativeImage::singlePixelSolidColor const):
(WebCore::NativeImage::draw):
(WebCore::NativeImage::clearSubimages):
* Source/WebCore/platform/graphics/skia/PathSkia.cpp: Added.
(WebCore::PathSkia::create):
(WebCore::PathSkia::PathSkia):
(WebCore::PathSkia::copy const):
(WebCore::PathSkia::platformPath const):
(WebCore::PathSkia::add):
(WebCore::PathSkia::addPath):
(WebCore::PathSkia::applySegments const):
(WebCore::PathSkia::applyElements const):
(WebCore::PathSkia::isEmpty const):
(WebCore::PathSkia::currentPoint const):
(WebCore::PathSkia::transform):
(WebCore::PathSkia::contains const):
(WebCore::PathSkia::strokeContains const):
(WebCore::PathSkia::fastBoundingRect const):
(WebCore::PathSkia::boundingRect const):
(WebCore::PathSkia::strokeBoundingRect const):
* Source/WebCore/platform/graphics/skia/PathSkia.h: Added.
(isType):
* Source/WebCore/platform/graphics/skia/PatternSkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::Pattern::createPlatformPattern const):
* Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::PlatformDisplay::skiaGLContext):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFont.cpp: Added.
(WebCore::SkiaHarfBuzzFont::getOrCreate):
(WebCore::harfBuzzFontFunctions):
(WebCore::createHarfBuzzFace):
(WebCore::SkiaHarfBuzzFont::SkiaHarfBuzzFont):
(WebCore::SkiaHarfBuzzFont::~SkiaHarfBuzzFont):
(WebCore::floatToHarfBuzzPosition):
(WebCore::doubleToHarfBuzzPosition):
(WebCore::SkiaHarfBuzzFont::scaledFont):
(WebCore::SkiaHarfBuzzFont::glyph):
(WebCore::SkiaHarfBuzzFont::glyphWidth):
(WebCore::SkiaHarfBuzzFont::glyphWidths):
(WebCore::SkiaHarfBuzzFont::glyphExtents):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFont.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFontCache.cpp: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
(WebCore::SkiaHarfBuzzFontCache::font):
(WebCore::SkiaHarfBuzzFontCache::remove):
(WebCore::SkiaHarfBuzzFontCache::clear):
* Source/WebCore/platform/graphics/skia/SkiaHarfBuzzFontCache.h: Copied from Source/WebCore/platform/graphics/PlatformPath.h.
* Source/WebCore/platform/graphics/skia/TransformationMatrixSkia.cpp: Added.
(WebCore::TransformationMatrix::TransformationMatrix):
(WebCore::TransformationMatrix::operator SkM44 const):
(WebCore::AffineTransform::AffineTransform):
(WebCore::AffineTransform::operator SkMatrix const):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::updateContents):
(WebCore::BitmapTexture::copyFromExternalTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawNumber):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerCairo.cpp: Added.
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp: Added.
(WebCore::AcceleratedBufferPool::singleton):
(WebCore::AcceleratedBufferPool::AcceleratedBufferPool):
(WebCore::AcceleratedBufferPool::~AcceleratedBufferPool):
(WebCore::AcceleratedBufferPool::acquireBuffer):
(WebCore::AcceleratedBufferPool::Entry::Entry):
(WebCore::AcceleratedBufferPool::Entry::markIsInUse):
(WebCore::AcceleratedBufferPool::Entry::canBeReleased const):
(WebCore::AcceleratedBufferPool::createAcceleratedBuffer):
(WebCore::AcceleratedBufferPool::scheduleReleaseUnusedBuffers):
(WebCore::AcceleratedBufferPool::releaseUnusedBuffersTimerFired):
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebCore/platform/graphics/texmap/coordinated/SurfaceUpdateInfo.h: Removed.
* Source/WebCore/platform/graphics/texmap/coordinated/Tile.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/TiledBackingStoreClient.h:
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp: Copied from Source/WebCore/platform/graphics/ImageBufferPlatformBackend.h.
(WebCore::ImageBackingStore::image const):
* Source/WebCore/platform/skia/SharedBufferSkia.cpp: Added.
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create):
(WebCore::SharedBuffer::createSkData const):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/API/c/cairo/WKImageCairo.cpp:
(WKImageCreateCairoSurface):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::updateBackingStore):
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebKit/Shared/ShareableBitmap.h:
(WebKit::ShareableBitmapConfiguration::imageInfo const):
* Source/WebKit/Shared/skia/ShareableBitmapSkia.cpp: Added.
(WebKit::ShareableBitmapConfiguration::validateColorSpace):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerRow):
(WebKit::ShareableBitmap::createGraphicsContext):
(WebKit::ShareableBitmap::paint):
(WebKit::ShareableBitmap::createImage):
(WebKit::ShareableBitmap::setOwnershipOfMemory):
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp: Copied from Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.h.
(IPC::ArgumentCoder<WebCore::Font>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::Font>::decodePlatformData):
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::decodePlatformData):
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::setCursor):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::IconCache::iconPath):
(WebKit::IconCache::iconBytes):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::CompositingCoordinator):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.h:
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
* Source/cmake/OptionsWPE.cmake:
* Tools/ImageDiff/Cairo.cmake:
* Tools/ImageDiff/PlatformImage.h:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewWindowProperties):
* Tools/TestWebKitAPI/glib/PlatformWPE.cmake:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.h:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/WebKitTestRunner/PlatformWebView.h:
* Tools/WebKitTestRunner/cairo/TestInvocationCairo.cpp:
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClient.h:
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.cpp:
(WTR::PlatformWebViewClientLibWPE::snapshot):
* Tools/WebKitTestRunner/libwpe/PlatformWebViewClientLibWPE.h:
* Tools/WebKitTestRunner/skia/TestInvocationSkia.cpp: Added.
(WTR::computeSHA1HashStringForPixmap):
(WTR::dumpPixmap):
(WTR::TestInvocation::dumpPixelsAndCompareWithExpected):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::snapshot):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.h:
* Tools/WebKitTestRunner/wpe/TestControllerWPE.cpp:
(WTR::TestController::takeViewPortSnapshot):
* Tools/wpe/backends/CMakeLists.txt:
* Tools/wpe/backends/HeadlessViewBackend.h:
* Tools/wpe/backends/fdo/HeadlessViewBackendFdo.cpp:
(WPEToolingBackends::HeadlessViewBackend::HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::~HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::snapshot):
(WPEToolingBackends::HeadlessViewBackend::updateSnapshot):

Canonical link: https://commits.webkit.org/274457@main
